core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java

```diff
@@ -67,6 +67,7 @@
 /**
  * Default {@link ClientRequestContext} implementation.
  */
+@UnstableApi
```

Comment:
Just want to confirm, I think it's possible to hide some of these completely, is the idea to mark as unstable first and then hide later?

Reply:
Is is possible to hide this class? Any user who wants to write a custom client implementation will need to use this API.

Reply:
If I'm not mistaken, only the constructor is ever needed. So we should be able to hide it by providing a factory in `*RequestContext`.

This class is literally what prompted me to file the issue about over-exposed APIs ;)

Reply:
I'm not sure if it's a good idea to have a static factory method with such large number of parameters. Perhaps it's better as it is now, i.e. public class annotated with `@UnstableApi`? It's for advanced users only anyway.

Reply:
I think it's still worth it. Though if we can merge the semantics with our current builders maybe even better not sure it's possible.

These classes leak a lot of implementation details, including a whole class `TimeoutController`. I think they're so core that we want to hide them as much as possible.

Reply:
Just realized that `setTimeoutController` and `TimeoutController` can be package-private. I think we can hide `DefaultClientRequestContext` if we do something about `DefaultClientRequestContext.init(EndpointGroup)`. Not sure if it makes sense to move it to `ClientRequestContext`..

Reply:
Let me keep thinking about how we can solve this. Meanwhile, let me leave this as it is and address it in another PR. At least `TimeoutController` is now removed from public API.